python-checks/src/main/java/org/sonar/python/checks/TfFunctionDependOnOutsideVariableCheck.java

```diff
@@ -55,29 +58,54 @@ private static void checkFunction(SubscriptionContext context) {
     }
     NameCollector collector = new NameCollector();
     functionDef.body().accept(collector);
-    Set<Symbol> b = new HashSet<>(collector.symbolToNames.keySet());
-    b.removeAll(functionDef.localVariables());
-    if (b.isEmpty()) {
+    Set<Symbol> allNames = new HashSet<>(collector.symbolToNames.keySet());
+    allNames.removeAll(functionDef.localVariables());
+    if (allNames.isEmpty()) {
       return;
     }
     var issue = context.addIssue(functionDef.name(), MESSAGE);
-    b.forEach(symbol -> collector.symbolToNames.get(symbol).forEach(name -> issue.secondary(name, MESSAGE_SECONDARY)));
+    allNames.forEach(symbol -> collector.symbolToNames.get(symbol).forEach(name -> issue.secondary(name, MESSAGE_SECONDARY)));
   }
 
   private static class NameCollector extends BaseTreeVisitor {
     Map<Symbol, List<Name>> symbolToNames = new HashMap<>();
 
     @Override
     public void visitName(Name pyNameTree) {
-      if (pyNameTree.isVariable()) {
-        Optional.ofNullable(pyNameTree.symbol())
-          .filter(symbol -> !symbol.is(Symbol.Kind.FUNCTION, Symbol.Kind.CLASS))
-          .filter(symbol -> symbol.usages().stream().map(Usage::kind).noneMatch(usageKind -> usageKind == Usage.Kind.IMPORT))
-          .ifPresent(symbol -> {
-            symbolToNames.putIfAbsent(symbol, new ArrayList<>());
-            symbolToNames.get(symbol).add(pyNameTree);
-          });
+      if (!pyNameTree.isVariable()) {
+        return;
       }
+      if (pyNameTree.type().mustBeOrExtend("tensorflow.Variable")) {
+        return;
+      }
+      if (isInstantiatedByTensorflow(pyNameTree)) {
+        return;
+      }
+
+      Optional.ofNullable(pyNameTree.symbol())
+        .filter(symbol -> !symbol.is(Symbol.Kind.FUNCTION, Symbol.Kind.CLASS, Symbol.Kind.AMBIGUOUS))
+        .filter(symbol -> symbol.usages().stream().map(Usage::kind).noneMatch(usageKind -> usageKind == Usage.Kind.IMPORT))
+        .ifPresent(symbol -> {
+          symbolToNames.putIfAbsent(symbol, new ArrayList<>());
+          symbolToNames.get(symbol).add(pyNameTree);
+        });
+    }
+
+    private static boolean isInstantiatedByTensorflow(Name pyNameTree) {
+      Symbol symbol = pyNameTree.symbol();
+      return symbol != null && symbol.usages().stream().filter(u -> u.kind().equals(Usage.Kind.ASSIGNMENT_LHS))
+        .map(Usage::tree)
+        .map(tree -> TreeUtils.firstAncestorOfKind(tree, Tree.Kind.ASSIGNMENT_STMT))
+        .filter(Objects::nonNull)
+        .map(AssignmentStatement.class::cast)
+        .map(AssignmentStatement::assignedValue)
+        .filter(assignedValue -> assignedValue.is(Tree.Kind.CALL_EXPR))
+        .map(CallExpression.class::cast)
+        .map(CallExpression::calleeSymbol)
+        .filter(Objects::nonNull)
+        .map(Symbol::fullyQualifiedName)
+        .filter(Objects::nonNull)
+        .anyMatch(fqn -> fqn.startsWith("tensorflow."));
```

Comment:
I think we generally should avoid those long streams that do different things at once (here, there are clearly two parts - searching for assigned values and checking whether they are Tensorflow values - this could be split).

For instance, I think the highlighted bit could be factored into a `isTensorflowValue`, which would then be easier to reason about in terms of what we consider to be a tensorflow object.

With a dedicated method, it might then be extended to consider more expressions (the one from the ruling could be handled there, and we even have an `Expressions::removeParentheses` helper that could help here).